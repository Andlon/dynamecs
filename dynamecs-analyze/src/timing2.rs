use std::collections::{BTreeMap, HashMap};
use std::iter::Peekable;
use std::time::Duration;
use eyre::eyre;
use time::OffsetDateTime;
use crate::{Record, RecordKind, Span, SpanPath, SpanTree};
use crate::RecordKind::{SpanClose, SpanNew};

// pub struct TimingTreeNode {
//     span: Span,
//     duration: Duration,
//     children: Vec<TimingTreeNode>,
// }
//
// pub struct StepTimings {
//     steps_trees: Vec<BTreeMap<String, TimingTreeNode>>,
//     step_index: usize,
// }
//
// pub struct StepTimingsCollection {
//     per_thread_timings: BTreeMap<String, TimingTreeNode>,
//     steps: Vec<TimingTreeNode>,
// }

// struct StepTimingAccumulator {
//     // a stack of span names coupled with timestamp of its *new* event
//     stack: Vec<(String, OffsetDateTime)>,
// }
//
// impl StepTimingAccumulator {
//     pub fn new() -> Self {
//         Self { stack: Vec::new() }
//     }
//
//     pub fn new_span(&mut self, span: &Span, new_timestamp: &OffsetDateTime) {
//         self.stack.push((span.name.clone(), *new_timestamp));
//     }
//
//     pub fn close_span(&mut self, span: &Span, close_timestamp: &OffsetDateTime) -> eyre::Result<()> {
//         // TODO: Proper errors
//         let (popped_span, new_timestamp) = self.stack.pop().ok_or_else(|| eyre!("tried to close span that was not opened"))?;
//         if span.name() != popped_span {
//             return Err(eyre!("unexpected span closed"));
//         }
//         let duration = *close_timestamp - new_timestamp;
//
//         Ok(())
//     }
// }

// pub struct StepTimings {
//     steps: Vec<()>,
//
// }

pub type TimingTree = SpanTree<Duration>;

#[derive(Debug, Clone)]
pub struct AccumulatedTimings {
    span_durations: HashMap<SpanPath, Duration>,
}

impl AccumulatedTimings {
    pub fn create_timing_tree(&self) -> TimingTree {
        let (paths, durations) = self.span_durations
            .iter()
            .map(|(path, duration)| (path.clone(), duration.clone()))
            .unzip();
        TimingTree::from_paths_and_payloads(paths, durations)
    }
}

#[derive(Debug, Clone)]
pub struct AccumulatedTimingSeries {
    steps: Vec<AccumulatedTimings>,
    // TODO: Timing from other sources outside of steps?
}

impl AccumulatedTimingSeries {
    pub fn steps(&self) -> &[AccumulatedTimings] {
        &self.steps
    }
}

pub fn extract_step_timings<'a>(records: impl IntoIterator<Item=&'a Record>) -> eyre::Result<AccumulatedTimingSeries> {
    // TODO: Collect statistics from spans outside run as well
    find_and_visit_dynamecs_run_span(records.into_iter())
}

fn find_and_visit_dynamecs_run_span<'a>(mut records: impl Iterator<Item=&'a Record>) -> eyre::Result<AccumulatedTimingSeries> {
    // First try to find the `run` span in the records
    while let Some(record) = records.next() {
        if let Some(span) = record.span() {
            if span.name() == "run"
                && record.target() == "dynamecs_app"
                && record.kind() == RecordKind::SpanNew {
                return visit_dynamecs_run_span(record, records);
            }
        }
    }

    Err(eyre!("Could not find new event for `run` span of dynamecs among records"))
}

fn visit_dynamecs_run_span<'a>(run_new_record: &Record, remaining_records: impl Iterator<Item=&'a Record>) -> eyre::Result<AccumulatedTimingSeries> {
    let run_thread = run_new_record.thread_id();
    let run_span_path = run_new_record.span_path();
    let mut iter = remaining_records;

    let mut steps = Vec::new();

    while let Some(record) = iter.next() {
        if record.thread_id() == run_thread {
            if let Some(span) = record.span() {
                match (span.name(), record.target(), record.kind()) {
                    ("step", "dynamecs_app", SpanNew) => {
                        if let Some(step) = visit_dynamecs_step_span(record, &mut iter)? {
                            // Only collect complete time steps
                            steps.push(step);
                        }
                    },
                    ("run", "dynamecs_app", SpanClose) if record.span_path() == run_span_path => {
                        break;
                    },
                    // TODO: Still accumulate timings for other things?
                    _ => {}
                }
            }
        }
    }

    Ok(AccumulatedTimingSeries {
        steps
    })
}

/// Returns accumulated timings for the next *complete* step in the records.
fn visit_dynamecs_step_span<'a>(
    step_new_record: &Record,
    remaining_records: &mut impl Iterator<Item=&'a Record>
) -> eyre::Result<Option<AccumulatedTimings>> {
    let step_path = step_new_record.span_path();

    let mut accumulator = TimingAccumulator::new();
    accumulator.new_span(step_path.clone(), step_new_record.timestamp().clone())?;

    while let Some(record) = remaining_records.next() {
        if record.thread_id() == step_new_record.thread_id() {
            if let Some(span) = record.span() {
                match record.kind() {
                    SpanNew => accumulator.new_span(record.span_path(),
                                                    record.timestamp().clone())?,
                    SpanClose => {
                        // TODO: use a stack to verify that open/close events are consistent?
                        let mut span_path = record.span_path();
                        // Close events don't report the current span anymore,
                        // so we need to add this to get a path consistent with the
                        // "new" event
                        span_path.push_span_name(span.name().to_string());
                        let is_step_span_path = span_path == step_path;
                        accumulator.close_span(span_path,
                                               record.timestamp().clone())?;
                        if span.name() == "step"
                            && record.target() == "dynamecs_app"
                            && is_step_span_path {
                            break;
                        }
                    },
                    _ => {}
                }
            }
        }
    }

    if accumulator.has_active_spans() {
        // If there are active spans, then the step is not yet complete,
        // so we do not want to include it in accumulation
        // (would lead to inconsistent time between parent and children)
        Ok(None)
    } else {
        Ok(Some(AccumulatedTimings { span_durations: accumulator.collect_completed_timings() }))
    }
}

#[derive(Debug)]
struct TimingAccumulator {
    completed_durations: HashMap<SpanPath, Duration>,
    timestamps_open: HashMap<SpanPath, OffsetDateTime>,
}

impl TimingAccumulator {
    pub fn new() -> Self {
        Self { completed_durations: Default::default(), timestamps_open: Default::default() }
    }

    pub fn new_span(&mut self, path: SpanPath, timestamp: OffsetDateTime) -> eyre::Result<()> {
        dbg!("new", &path);
        if self.timestamps_open.insert(path, timestamp).is_some() {
            return Err(eyre!("tried to create new span that is already active (not closed)"));
        }
        Ok(())
    }

    pub fn close_span(&mut self, path: SpanPath, timestamp_close: OffsetDateTime) -> eyre::Result<()> {
        dbg!("close", &path);
        let timestamp_new = self.timestamps_open.remove(&path)
            .ok_or_else(|| eyre!("found close event for span that is not currently active. Span path: {path}"))?;
        let span_duration: Duration = (timestamp_close - timestamp_new).unsigned_abs();
        let mut accumulated_duration = self.completed_durations.entry(path)
            .or_default();
        *accumulated_duration += span_duration;
        Ok(())
    }

    pub fn has_active_spans(&self) -> bool {
        !self.timestamps_open.is_empty()
    }

    pub fn collect_completed_timings(self) -> HashMap<SpanPath, Duration> {
        self.completed_durations
    }
}

