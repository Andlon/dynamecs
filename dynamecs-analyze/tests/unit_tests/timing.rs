use std::path::PathBuf;
use dynamecs_analyze::iterate_records;
use dynamecs_analyze::timing2::{extract_step_timings, format_timing_tree};

#[test]
fn extract_timings() -> eyre::Result<()> {
    // TODO: Modify iterate_records so that we don't have to collect first
    // (probably remove the internal Result<Record> layer
    let path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests/unit_tests/example_logs/basic_app_3_steps.jsonlog");
    let records: Vec<_> = iterate_records(path)?
        .collect::<eyre::Result<Vec<_>>>()?;
    let timings = extract_step_timings(records).unwrap();

    let trees: Vec<_> = timings.steps().iter().map(|step| step.timings.create_timing_tree()).collect();

    // First step
    {
        let tree = &trees[0];
        dbg!(&tree);
        println!("{}", format_timing_tree(&tree));
    }

    Ok(())
}