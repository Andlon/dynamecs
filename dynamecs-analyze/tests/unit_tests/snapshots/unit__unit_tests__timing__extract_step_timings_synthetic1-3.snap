---
source: dynamecs-analyze/tests/unit_tests/timing.rs
expression: format_timing_tree(&summary)
---
Total     Average   Self     Count  Rel parent  Rel root  Span                          
═════════════════════════════════════════════════════════════════════════════════════════
 25.0 s    25.0 s     8.0 %      1         N/A  100.0 %   run                           
  0.0 s     0.0 s       N/A      1       0.0 %    0.0 %   ├── init                      
 23.0 s    11.5 s    21.7 %      2      92.0 %   92.0 %   └── step                      
 18.0 s     9.0 s    11.1 %      2      78.3 %   72.0 %       └── simulate              
  8.0 s     2.7 s   100.0 %      3      44.4 %   32.0 %           ├── assemble          
  4.0 s     4.0 s   100.0 %      1      22.2 %   16.0 %           ├── occasional        
  4.0 s     2.0 s   100.0 %      2      22.2 %   16.0 %           └── solve             
═════════════════════════════════════════════════════════════════════════════════════════

